### ISSUE TITLE
Table with name _ does not exist when using duckdb.query on a pandas dataframe in IPython unless wrapped in print or .to_df()

### ISSUE THREAD

**Issue by davidwales** [duckdb/duckdb#2966]
#### What happens?

Suppose I have a pandas dataframe called `mydf` and I'm _using IPython_.
If I evaluate a query directly, using `duckdb.query` I get:

```
---------------------
-- Expression Tree --
---------------------
Subquery [select sum(a) from mydf]

---------------------
-- Result Columns  --
---------------------
- sum(a) (HUGEINT)

---------------------
-- Result Preview  --
---------------------
Catalog Error: Table with name mydf does not exist!
```
If I do exactly the same thing in a standard Python repl instead of IPython, it works:

```
---------------------
-- Result Preview  --
---------------------
sum(a)
HUGEINT
[ Rows: 1]
6
```

Alternatively, if I stay in IPython and wrap the call to `duckdb.query` inside `print()`, or if I append `.to_df()`, then it also works!

#### To Reproduce

**Run this in IPython**. It works without issue in the default Python interpreter.

``` python
### In IPython ###
import pandas as pd
import duckdb

mydf = pd.DataFrame({'a': [1, 2, 3]})
duckdb.query("select sum(a) from mydf")
# Catalog Error: Table with name mydf does not exist!

print(duckdb.query("select sum(a) from mydf"))
# ---------------------
# -- Result Preview  --
# ---------------------
# sum(a)
# HUGEINT
# [ Rows: 1]
# 6

duckdb.query("select sum(a) from mydf").to_df()
#    sum(a)
# 0     6.0
```

**EDIT**: IPython uses the [display](https://ipython.readthedocs.io/en/stable/api/generated/IPython.display.html#IPython.display.display) function to render Python objects in the repl. You can replicate this issue in a standard Python repl as follows:

``` python
### In standard Python repl ###
import pandas as pd
import duckdb
from IPython.display import display

mydf = pd.DataFrame({'a': [1, 2, 3]})
display(duckdb.query("select sum(a) from mydf"))
# Catalog Error: Table with name mydf does not exist!
```

#### Environment (please complete the following information):
 - OS: Windows 10
 - DuckDB Version: 0.3.1
 - DuckDB Client: IPython 7.30.1 with Python 3.8.7

#### Before Submitting

- [x] **Have you tried this on the latest `master` branch?**
* **Python**: `pip install duckdb --upgrade --pre`
* **R**: `install.packages("https://github.com/duckdb/duckdb/releases/download/master-builds/duckdb_r_src.tar.gz", repos = NULL)`
* **Other Platforms**: You can find binaries [here](https://github.com/duckdb/duckdb/releases/tag/master-builds) or compile from source.

- [x] **Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?**

**Comment by pdet** [comment]
Hm,
What you are using, in this case, is the replacement scans. The replacement scans work by searching for a python object in the local or global space from the function call.

I'm not super familiar with Ipython,  so please correct me if I'm wrong. It seems that when the display method is called, the IPython enviroment loses track of the local and global variables of the original script. Since these variables can't be accessed when we actually execute the query, inside the display method, it shouts the table not found error.

A workaround for this is to use the ```con.register(mydf)``` method.

You can simulate it by debugging the script and running ```locals()``` and ```globals()```, if you step into the display method, the name of the variables, and part of them seem to be lost.

``` python
import pandas as pd
import duckdb
from IPython.display import display
import pdb
pdb.set_trace()

mydf = pd.DataFrame({'a': [1, 2, 3]})
mydf_2 = pd.DataFrame({'a': [2, 3,4]})

con = duckdb.connect()

display(con.query("select sum(mydf.a) from mydf inner join mydf_2 on (mydf.a = mydf_2.a)"))
```

**Comment by Alex-Monahan** [comment]
As a note, if there is a way to allow DuckDB to see Pandas df's from iPython, that would likely allow Jupyter notebooks to use the %%sql magic from iPython-SQL! 
[iPython SQL issue](https://github.com/catherinedevlin/ipython-sql/issues/199)
[DuckDB Engine issue](https://github.com/Mause/duckdb_engine/issues/136)

**Comment by davidwales** [comment]
@Alex-Monahan In fact, that's how I discovered this bug! I was trying to combine your [IPython SQL tutorial](https://alex-monahan.github.io/2021/08/22/Python_and_SQL_Better_Together.html) with the [DuckDB SQL on Pandas article](https://duckdb.org/2021/05/14/sql-on-pandas.html).

**Comment by Alex-Monahan** [comment]
> Hey @Alex-Monahan is this only related to sqlalchemy?

Hmm, it's tough to say where the issue lies. I'm not sure if a change in duckdb-engine would fix things, or if duckdb needs its cursors to look more like the Python DB API standard. But a regular call to duckdb-engine /SQLAlchemy works(example is in the notebook), it's just how iPython-SQL uses SQLAlchemy. Does that help? It's an interaction of all 3 - iPython-SQL, duckdb-engine, and DuckDB's Python client.

**Comment by davidwales** [comment]
I just tried to test this by installing with `pip install duckdb --upgrade --pre`. However, I'm not sure I'm actually getting the latest version, as I can see that the CI action for this merge was cancelled (at least for the Windows Py 3.8 build).

`duckdb.__version__` reports: `0.3.2-dev1201`

**Comment by Mause** [comment]
@Alex-Monahan that rowcount issue was a bug in duckdb_engine, I've fixed it in https://github.com/Mause/duckdb_engine/pull/217

**Comment by Alex-Monahan** [comment]
> @Alex-Monahan that rowcount issue was a bug in duckdb_engine, I've fixed it in [Mause/duckdb_engine#217](https://github.com/Mause/duckdb_engine/pull/217)

Thank you! I'll test that out today or tomorrow!

**Comment by Alex-Monahan** [comment]
> @Alex-Monahan that rowcount issue was a bug in duckdb_engine, I've fixed it in [Mause/duckdb_engine#217](https://github.com/Mause/duckdb_engine/pull/217)

Working like a charm again. Much appreciated!

**Comment by Alex-Monahan** [comment]
I was able to test this out, and iPython-SQL and DuckDB engine are now able to detect Pandas dataframes! This is fantastic! Thanks for solving this - it makes the workflow really smooth now.
